ksqldb-engine/src/main/java/io/confluent/ksql/structured/SchemaKSourceFactory.java

```diff
@@ -174,15 +182,34 @@ private SchemaKSourceFactory() {
 
     final SourceStep<KTableHolder<GenericKey>> step;
 
-    if (buildContext.getKsqlConfig().getBoolean(KsqlConfig.KSQL_ROWPARTITION_ROWOFFSET_ENABLED)) {
+    final int pseudoColumnVersionToUse = determinePseudoColumnVersionToUse(buildContext);
+
+    // If the old query has a v1 table step, continue to use it.
+    // See https://github.com/confluentinc/ksql/pull/7990
+    boolean useOldExecutionStepVersion = false;
+    if (buildContext.getPlanInfo().isPresent()) {
+      final Set<ExecutionStep<?>> sourceSteps = buildContext.getPlanInfo().get().getSourceSet();
+      useOldExecutionStepVersion = sourceSteps
+          .stream()
+          .anyMatch(executionStep -> executionStep instanceof TableSourceV1);
+    }
+
+    if (useOldExecutionStepVersion && pseudoColumnVersionToUse != 0) {
+      throw new IllegalStateException("TableSourceV2 was released in conjunction with pseudocolumn"
+          + "version 1. Something has gone very wrong");
+    }
+
+    // When feature flag is removed, remove getBoolean() from condition
```

Comment:
nit: I don't think this comment is necessary. If `KsqlConfig.KSQL_ROWPARTITION_ROWOFFSET_ENABLED` is deleted, then this line will have to be updated (else compilation will fail) and it's pretty clear how this should be updated when that happens.